sourcegraph/cody
# Add multi-repo search for enterprise, remove remote embeddings

- Removes remote embeddings from consumer and enterprise
- Adds remote search through the `getCodyContext` GraphQL endpoint as a new context provider for Enterprise
- The current file's repository (if any), or `cody.codebase`, is automatically included by this context source
- Adds a multi-repo picker to Enterprise chats; up to 9 manually selected remote repos are saved in the chat transcript
- New chats default to using the workspace roots' repositories for context

Known issues:
- Needs automated tests for VSCode

Fixes #2622, #2624, #2850

## Test plan

This change deletes a lot of unused code so naturally, run the unit tests, e2e tests, etc.

Manual test plan:

Remote embeddings have been removed for consumer users:

1. Clone a popular open source repository, such as `git@github.com/react/react.git`
2. Open VSCode and sign in to a consumer account
3. Start a chat
4. Verify that the enhanced context selector advertises Search (that is, symf) and Embeddings with "set up embeddings." You should *not* see "sourcegraph.com" (that is, remote embeddings.)
5. Open a file in the repository, Cmd-Shift-P, Cody Command: Edit Code and edit the code. The edit should succeed.

Enterprise changes, so *sign into an Enterprise account*, then:

Repo selection:

1. Open a new VSCode session with nothing in the workspace.
2. Start a chat.
3. Open the Enhanced Context Selector.
4. Verify that you see a message about no repositories.
5. Click Choose Repositories.
6. Verify the Repository Picker works: that you can pick up to 9 repositories manually, that you can add and remove repositories, that you can search repositories, that ESC or a click outside the picker does not affect the repositories chosen, etc.
7. Verify that the Enhanced Context Selector works: Mouse over the repository names and see the full name; click X to remove a repository; etc.

Chats use the selected repositories; chats save the selected repositories:

1. Start a chat and add some repositories.
2. Chat with Cody and verify that context is fetched from the selected repositories.
3. Close the chat.
4. Open a new chat and verify that the repository list is empty again.
5. Reopen the first chat and verify that the repository list are the ones selected in step 2.

Chats automatically use the repository workspace roots for context:

1. Open a clone of a repository indexed by the enterprise.
2. Start a chat.
3. Verify that the Enhanced Context Selector includes the relevant repository. (If this is a different repository, you may have `cody.codebase` set.)

Supports multiple workspace roots:

1. File, Add Folder to Workspace, select another folder which is a clone of a repository indexed by the enterprise.
2. Start a chat.
3. Verify that the Enhanced Context Selector includes all the workspace root repositories (max 9) and that you can manually remove them (expect the automatically included one.)
4. Open a file from the first folder in the workspace; verify the related repository is automatically included (has an "info" mark in the Enhanced Context Selector.)
5. Open a file from the second folder in the workspace; verify the related repository is automatically included (info mark.) Note, `cody.codebase` will override this automatic switching behavior. For example, if you use s2 and open the github.com/sourcegraph/sourcegraph repo the `.vscode/settings.json` file sets `"cody.codebase": "github.com/sourcegraph/sourcegraph"`. If your second folder is the github.com/sourcegraph/cody repository, you won't get automatic inclusion of the Cody repository by design.
6. Set the `cody.codebase` setting, start a new chat, and check the automatically included repository matches `cody.codebase`.

Inline edits:

1. Open a file.
2. Cmd-Shift-P, Cody Command: Edit Code
3. Instruct Cody and check that the edit was carried out
